spec/app/ics-070-shared-queue/README.md

```diff
@@ -0,0 +1,485 @@
+# Queue Module
+
+## Introduction
+
+This document presents a technical specification of the **queue module**.
+The queue module contains a logic necessary for implementing shared FIFO queue among two parties (i.e., two blockchains).
+Therefore, in an implementation of the FIFO queue we are interested in, the following entities take place:
+  - Two blockchains: Each validator of each blockchain implements its queue module.
+  We say that one blockchain is the parent blockchain, whereas the other is the baby blockchain.
+  - IBC communication: There exists an IBC communication among two aforementioned blockchains.
+  Blockchains communicate exclusively using the IBC channels among them.
+
+We present two versions of the queue module.
+The versions provide slightly different interface and guarantees, as we describe in the following subsection.
+
+### Shared FIFO Queue Specification
+
+Our shared FIFO queue is a concurrent object.
+We now define both versions of our queue.
+
+#### Sequential Consistent FIFO Queue
+
+The first "version" of the queue - which is denote by Sequential Consistent FIFO queue (SCQ) - exposes the following interface:
+- *enqueue(x)* operation: Operation that enqueues the element *x* to the queue.
+- *dequeue()* operation: Operation that dequeues the first element of the queue and returns that element.
+- *peak()* operation: Operation that returns the first element of the queue without removing it.
+- *contains(x)* operation: Operation that returns whether the element *x* belongs to the queue.
+
+We set the following constraints:
+- Enqueue and contains operations are invoked solely by the parent blockchain.
+- Dequeue and peak operations are invoked solely by the baby blockchain.
+
+Lastly, our implementation satisfies the **sequential consistency** correctness criterium:
+The result of any execution is the same as if the operations of all the processes were executed in some sequential order, and the operations of each individual process appear in this sequence in the order specified by its protocol.
+
+#### Causally Consistent FIFO Queue
```

Comment:
And this the proposed causally ordered channels. It would make sense to contribute to https://github.com/cosmos/ibc/issues/550 rather than invent new concepts.

Reply:
Thanks a lot for the pointer. As far as I see there are slight differences. #550 talks about explicitly specifying predecessors, while in causal consistency (on the interface) relations are not explicitly stated. Causal consistency is a well-establish correctness criteria for shared data structures in concurrent programming. Actually, from our current understanding in the two chain case, ordered channels coincide with this "Causally Consistent FIFO Queue", while in the multi-chain case, we would need additional indices.